src/GNDStk/gnds-utility.hpp

```diff
@@ -0,0 +1,231 @@
+   template<class...> class CCONTO
+>
+bool convert(
+   const gnds::Tree<MCONFROM,CCONFROM> &,
+   gnds::Tree<MCONTO,CCONTO> &
+);
+
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(
+   const gnds::Tree<MCON,CCON> &,
+   gnds::xml &
+);
+
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(
+   const gnds::Tree<MCON,CCON> &,
+   gnds::json &
+);
+
+
+// xml to {Tree,xml,json}
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(const gnds::xml &, gnds::Tree<MCON,CCON> &);
+
+bool convert(const gnds::xml &, gnds::xml &);
+
+bool convert(const gnds::xml &, gnds::json &);
+
+
+// json to {Tree,xml,json}
+template<
+   template<class...> class MCON,
+   template<class...> class CCON
+>
+bool convert(const gnds::json &, gnds::Tree<MCON,CCON> &);
+
+bool convert(const gnds::json &, gnds::xml &);
+
+bool convert(const gnds::json &, gnds::json &);
+
+
+
+// -----------------------------------------------------------------------------
+// Some string constants
+// -----------------------------------------------------------------------------
+
+namespace detail {
+
+// For pugi::xml_node_types { node_pcdata, node_cdata, node_comment }
+inline const std::string keyword_pcdata  = "body";
+inline const std::string keyword_cdata   = "text";
+inline const std::string keyword_comment = "comment";
+
+// Aliases
+inline const std::string &keyword_body = keyword_pcdata;
+inline const std::string &keyword_text = keyword_cdata;
+
+} // namespace detail
+
+
+
+// -----------------------------------------------------------------------------
+// Utility constructs
+// The functions here could possibly go into the detail namespace, but could
+// arguably be useful, in their own right, to users. So, I'll leave them out
+// in the overall project namespace (which enclosed the #include of this file).
+// -----------------------------------------------------------------------------
+
+// filesize(char *)
+inline std::ifstream::pos_type filesize(const char * const file)
+{
+   std::ifstream ifs(file, std::ifstream::ate | std::ifstream::binary);
+   return ifs.tellg();
+}
+
+// filesize(string)
+inline std::ifstream::pos_type filesize(const std::string &file)
+{
+   return filesize(file.c_str());
+}
+
+
+// endsin
+// C++20 will have an ends_with(); for now we'll have this
+inline bool endsin(const std::string &str, const std::string &end)
+{
+   return str.size() >= end.size() && &str[str.size()-end.size()] == end;
+}
+
+// typeof
+#ifdef GNDS_BOOST
+template<class T>
+inline void typeof(const T &)
+{
+   std::cout << boost::core::demangle(typeid(T).name()) << std::endl;
+}
+#endif
+
+
+
+// -----------------------------------------------------------------------------
+// Re: file extensions
+// -----------------------------------------------------------------------------
+
+// xml
+inline bool endsin_xml(const std::string &str)
```

Comment:
Is this really needed? I'm not comfortable relying on an extension to tell me what kind of file I have.

Reply:
Actually I'm not using it for that purpose; I'm using the file magic number. The extension, if any, is used for a different purpose; I can explain the exact details of what's going on another time. For writing, the file magic number is either unavailable or meaningless, so I do use the extension (unless the user has explicitly requested a format) to decide what format is wanted for the output file.